runtime/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tokio Runtime wrapper.
+
+use std::{fmt, thread};
+use std::sync::mpsc;
+use std::time::{Duration, Instant};
+use futures::{future, Future, IntoFuture};
+pub use tokio::timer::Delay;
+pub use tokio::runtime::{Runtime as TokioRuntime, Builder as TokioRuntimeBuilder, TaskExecutor};
+
+/// Runtime for futures.
+///
+/// Runs in a separate thread.
+pub struct Runtime {
+	executor: Executor,
+	handle: RuntimeHandle,
+}
+
+impl Runtime {
+	fn new(runtime_bldr: &mut TokioRuntimeBuilder) -> Self {
+		let mut runtime = runtime_bldr
+			.build()
+			.expect("Building a Tokio runtime will only fail when mio components \
+				cannot be initialized (catastrophic)");
+		let (stop, stopped) = futures::oneshot();
+		let (tx, rx) = mpsc::channel();
+		let handle = thread::spawn(move || {
+			tx.send(runtime.executor()).expect("Rx is blocking upper thread.");
+			runtime.block_on(futures::empty().select(stopped).map(|_| ()).map_err(|_| ()))
+				.expect("Tokio runtime should not have unhandled errors.");
+		});
+		let executor = rx.recv().expect("tx is transfered to a newly spawned thread.");
+
+		Runtime {
+			executor: Executor {
+				inner: Mode::Tokio(executor),
+			},
+			handle: RuntimeHandle {
+				close: Some(stop),
+				handle: Some(handle),
+			},
+		}
+	}
+
+	/// Spawns a new tokio runtime with a default thread count on a background
+	/// thread and returns a `Runtime` which can be used to spawn tasks via
+	/// its executor.
+	pub fn with_default_thread_count() -> Self {
+		let mut runtime_bldr = TokioRuntimeBuilder::new();
+		Self::new(&mut runtime_bldr)
+	}
+
+	/// Spawns a new tokio runtime with a the specified thread count on a
+	/// background thread and returns a `Runtime` which can be used to spawn
+	/// tasks via its executor.
+	#[cfg(any(test, feature = "test-helpers"))]
+	pub fn with_thread_count(thread_count: usize) -> Self {
+		let mut runtime_bldr = TokioRuntimeBuilder::new();
+		runtime_bldr.core_threads(thread_count);
+
+		Self::new(&mut runtime_bldr)
+	}
+
+	/// Returns this runtime raw executor.
+	#[cfg(any(test, feature = "test-helpers"))]
+	pub fn raw_executor(&self) -> TaskExecutor {
+		if let Mode::Tokio(ref executor) = self.executor.inner {
+			executor.clone()
+		} else {
+			panic!("Runtime is not initialized in Tokio mode.")
+		}
+	}
+
+	/// Returns runtime executor.
+	pub fn executor(&self) -> Executor {
+		self.executor.clone()
+	}
+}
+
+#[derive(Clone)]
+enum Mode {
+	Tokio(TaskExecutor),
+	// Mode used in tests
+	#[allow(dead_code)]
+	Sync,
+	// Mode used in tests
+	#[allow(dead_code)]
```

Comment:
```suggestion
	#[cfg(any(test, feature = "test-helpers"))]
```

Reply:
This will lead to some ugly code further in the places, when these constants are used. I'll have to cover all such places with this condition

Reply:
Like this: https://github.com/paritytech/parity-common/pull/271/files#diff-11f21b7f145f98241270cf825e0aaab5R112

Reply:
fair enough